ee/src/main/java/org/jboss/as/ee/structure/GlobalDirectoryDependencyProcessor.java

```diff
@@ -0,0 +1,102 @@
+import org.jboss.as.server.deployment.DeploymentUnitProcessingException;
+import org.jboss.as.server.deployment.DeploymentUnitProcessor;
+import org.jboss.as.server.deployment.module.ModuleSpecification;
+import org.jboss.as.server.moduleservice.ExternalModule;
+import org.jboss.modules.ModuleLoader;
+import org.jboss.msc.service.ServiceBuilder;
+import org.jboss.msc.service.ServiceController;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.ServiceRegistry;
+import org.jboss.msc.service.ServiceTarget;
+
+/**
+ * A deployment processor that prepares the global directories that are going to be used as dependencies in a deployment.
+ * <p>
+ * It finds which services were installed under ee/global-directory resource using its capability name and installs a
+ * {@link GlobalDirectoryDeploymentService} to consume the supplied GlobalDirectory values.
+ * <p>
+ * This deployment processor ensures that the {@link GlobalDirectoryDeploymentService} installed for this deployment is up before moving
+ * to the next phase.
+ *
+ * @author Yeray Borges
+ */
+public class GlobalDirectoryDependencyProcessor implements DeploymentUnitProcessor {
+
+    @Override
+    public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitProcessingException {
+        final DeploymentUnit deploymentUnit = phaseContext.getDeploymentUnit();
+        final ServiceName depUnitServiceName = deploymentUnit.getServiceName();
+
+        final DeploymentUnit parent = deploymentUnit.getParent();
+        final DeploymentUnit topLevelDeployment = parent == null ? deploymentUnit : parent;
+        final ExternalModule externalModuleService = topLevelDeployment.getAttachment(Attachments.EXTERNAL_MODULE_SERVICE);
+        final ModuleLoader moduleLoader = deploymentUnit.getAttachment(Attachments.SERVICE_MODULE_LOADER);
+        final ModuleSpecification moduleSpecification = deploymentUnit.getAttachment(Attachments.MODULE_SPECIFICATION);
+        final CapabilityServiceSupport capabilitySupport = deploymentUnit.getAttachment(Attachments.CAPABILITY_SERVICE_SUPPORT);
+
+        final ServiceRegistry serviceRegistry = phaseContext.getServiceRegistry();
+        final ServiceTarget target = phaseContext.getServiceTarget();
+        final ServiceTarget externalServiceTarget = deploymentUnit.getAttachment(Attachments.EXTERNAL_SERVICE_TARGET);
+
+        final ServiceName allDirReadyServiceName = depUnitServiceName.append("directory-services-ready");
+        final ServiceBuilder<?> allDirReadyServiceBuilder = target.addService(allDirReadyServiceName);
+        final List<Supplier<GlobalDirectoryResourceDefinition.GlobalDirectory>> allDirReadySuppliers = new ArrayList<>();
+
+        List<ServiceName> serviceNames = serviceRegistry.getServiceNames();
+        for (ServiceName serviceName : serviceNames) {
+            if (capabilitySupport.getCapabilityServiceName(EE_GLOBAL_DIRECTORY_CAPABILITY_NAME).isParentOf(serviceName)) {
```

Comment:
Better is to call capabilitySupport.getCapabilityServiceName(EE_GLOBAL_DIRECTORY_CAPABILITY_NAME) outside the loop and cache the result in a local var.

Reply:
Move outside of the loop, thanks